File: Leetcode/Java/116. Populating Next Right Pointers in Each Node.java
public class Solution {
    public void connect(TreeLinkNode root) {
        TreeLinkNode p = root;
        while(p != null) {
            TreeLinkNode nextP = p.left;
            TreeLinkNode c = nextP;
            while(c != null) {
                if(p.left == c) {
                    c.next = p.right;
                } else {
                    if(p.next == null) {
                        c.next = null;
                    }
                    else {
                        c.next = p.next.left;
                    }
                    p = p.next;
                }
                c = c.next;
            }
            p = nextP;
        }
    }
}

File: Leetcode/Javascript/116. Populating Next Right Pointers in Each Node.js
/*
    Recursion
*/
var connect = function(root) {
    if(!root) return;
    var p = root.next;
    if(root.right && p) root.right.next = p.left;
    if(root.left) root.left.next = root.right;
    connect(root.right);
    connect(root.left);
};
/*
    Iteration
*/
var connect = function(root) {
    var ptr = root;
    while(ptr) {
        var nextPtr = ptr.left;//beginning point of next layer
        var child = nextPtr;
        while(child) {
            if(child == ptr.left) child.next = ptr.right;
            else {
                ptr = ptr.next;
                if(ptr) child.next = ptr.left;
            }
            child = child.next;
        }
        ptr = nextPtr;
    }
};
